### PR TITLE
Add sine model

### PR DESCRIPTION
**Pull request by lneuhaus**
#### Description
This PR adds a SineModel to models.py to fit sines. This was briefly discussed in https://groups.google.com/g/lmfit-py/c/uCF4tUT7DGs/m/YjtEuoJ1AwAJ

###### Type of Changes
- [ ] Bug fix
- [x] New feature
- [ ] Refactoring / maintenance
- [ ] Documentation / examples

Also adds a base class that might be useful to add tests for models.py.

###### Tested on
Python: 3.7.9 (default, Aug 31 2020, 12:42:55) 
[GCC 7.3.0]

lmfit: 0.9.15rc1+175.ged81916, scipy: 1.5.2, numpy: 1.19.1, asteval: 0.9.19, uncertainties: 3.1.4

###### Verification 
Have you
- [x] included docstrings that follow PEP 257?

- [x] referenced existing Issue and/or provided relevant link to mailing list?

- [x] verified that existing tests pass locally?

- [x] verified that the documentation builds locally?

- [ ] squashed/minimized your commits and written descriptive commit messages?

- [x] added or updated existing tests to cover the changes?
- [x] updated the documentation?
- [ ] added an example?

### CODE REVIEW

**Comment by newville** [comment]
@lneuhaus It appears that @reneeotten is gone.  I have a few comments and requests:

- The "all checks have failed" message is coming from Py3.5, which doesn't recognize the f-string in the test.   We intend to drop support for Py3.5 anyway, so I just removed this from the .travis.yml.  Can you rebase from the current master branch?

- The `BaseTestForModels` class looks useful, but it would odd to have it and use it for only one Model.   Especially in light of further comments, I would suggest shelving this approach for this PR about adding SineModel. If you would like to make another PR that implements `BaseTestForModels` and uses it for several Models, that would be welcome.

- None of the other model subclasses have `guess_and_fit()` or `post_process_fit_result()`.   I don't think we need `guess_and_fit()` - that just combines calls and it buries the parameters.

- The idea of `post_process_fit_result()` is intriguing and I can see that having such a "hook".  Like, one could argue that the whole "FWHM" and "Height" parameters of all the "peak" models are needed as calculated parameters *during* the fit, and could be determined in post-processing (there are counter-arguments for that -- the propagated uncertainties for example).  So, I'm not opposed to a `Model.post_fit()` method that did nothing by default but could be overridden by any subclass.    
But: I don't think you really need it here.   The much-abused and apparently much-misunderstood `param_hints` is exactly the right thing to use here.  Your `SineModel` can impose `amplitude>=0`, `frequency>=0` and `-pi < shift < pi` with:

    def _set_paramhints_prefix(self):
        self.set_param_hint('amplitude', min=0)
        self.set_param_hint('frequency', min=0)
        self.set_param_hint('phase', min=-np.pi, max=np.pi)

These will be satisfied with your `guess()` method,  if you change the domain of the shift_guesses to be:

     shift_guesses = np.linspace(-1, 1, 11)*np.pi

And then, you would not need any post-processing.

**Comment by lneuhaus** [comment]
Thanks for the comments, you must have guessed that I was fishing for feedback on how to use param-hints here. I'll get tests passing and clean things up, and will request your review.

**Comment by lneuhaus** [comment]
> self.set_param_hint('phase', min=-np.pi, max=np.pi)

Is the convention in this repository to have phases range predominantly from `-pi` to `+pi`, rather than `0` to `2*pi`?

**Comment by newville** [comment]
@lneuhaus Thanks!  No preference on [-pi, pi] vs [0, 2pi]  I guess we should all be encouraging the use of 
[tau](https://tauday.com/tau-manifesto), so  `[0, 1](2pi) is probably better.

**Comment by lneuhaus** [comment]
> No preference on [-pi, pi] vs [0, 2pi] I guess we should all be encouraging the use of tau, so `0, 1 is probably better.

I kept `[0, 2pi]` for now, as I would have to touch the sine function in `lineshapes.py` for the move towards `tau`, which could break things or dependencies I'm not sure I understand. Also, other functions there use ordinary radians (e.g. `expsine`). So I suggest to migrate to `tau` in a separate PR if one desires so.

**Comment by lneuhaus** [comment]
I'm quite happy with the fitting performance now, as the only test failures I get are due to too tight tolerances for too few data points / too much noise - in other words, I was unable to find any pathological cases the model could not fit. 

> squashed/minimized your commits and written descriptive commit messages?

I am not sure I understand what this means. Why not simply squash+merge this PR when it is time for that?

> The BaseTestForModels class looks useful, but it would odd to have it and use it for only one Model. Especially in light of further comments, I would suggest shelving this approach for this PR about adding SineModel. If you would like to make another PR that implements BaseTestForModels and uses it for several Models, that would be welcome.

I simply added a few more tests here to motivate not removing work that has already been done. If you'd prefer I would remove all that and make the test only cover the SineModel.

**Comment by lneuhaus** [comment]
@newville I was unable to request your review (permission settings I guess), so I'm just tagging you here.

**Comment by newville** [comment]
@lneuhaus I think this looks good.  I'm not very concerned about "squashed/minimized" PRs.  I don't mind seeing a little bit of the intermediate steps in getting to the final merge.

I think BaseTestForModels is fine to leave in - it is definitely a nice approach and we're more likely to use it now that the way has been paved, so to speak. So, thanks!

"tau" is sort of a joke, but awesome to include.

I do see that the docstring has a "+ B"  - should that be removed?

I'm happy to merge this soon.

File: lmfit/models.py [hunk 1]
```diff
@@ -327,6 +328,56 @@ def guess(self, data, x=None, **kwargs):
...
 
 
+class SineModel(Model):
+    r"""A model based on a sinusoidal lineshape.
+
+    The model has three Parameters: `amplitude`, `frequency`, and `shift`.
+
+    .. math::
+
+        f(x; A, \phi, f) = A \sin (f x + \phi) + B
```
Comment (lneuhaus) [inline lmfit/models.py:338]:
```suggestion
        f(x; A, \phi, f) = A \sin (f x + \phi) 
```

**Comment by lneuhaus** [comment]
> I do see that the docstring has a "+ B" - should that be removed?

Yes, that was a left-over, thanks. 

> I'm happy to merge this soon.

Awesome! Feel free to merge anytime (It seems as if CI takes incredibly long... not sure that is normal).

**Comment by newville** [comment]
@lneuhaus Great - will merge.  Yes, some of the examples in the CI scripts are pretty slow.  But it passed and only the docstring changed, so I'll merge.  Thanks!!

**Comment by reneeotten** [comment]
@lneuhaus @newville some of the tests added in this PR fail for me locally on macOS:
```
========================================================================================= short test summary info ==========================================================================================
FAILED test_models.py::TestSineModel::test_less_than_1_period[1] - AssertionError: wrong fit for parameter frequency: expected 0.65, fitted 0.9133860189041971.
FAILED test_models.py::TestSineModel::test_typical_regime[6] - AssertionError: wrong fit for parameter amplitude: expected 23.5, fitted 22.529682640337313.
================================================================================= 2 failed, 24 passed, 1 warning in 0.50s ==================================================================================
```

it does pass however on Travis CI so it could be something macOS specific (e.g., the use of `random` comes to mind here). 
@lneuhaus did you run the tests locally on Linux or on macOS (from `GCC 7.3.0` it's likely Linux)? Is anyone else seeing this? I would like to fix this so that the test-suite passes locally for me again.

**Comment by lneuhaus** [comment]
I ran tests on Ubuntu 20.04. Note that the random number generator is seeded in the test module, but I guess this might still result in a different random number sequence? The only thing that then comes to my mind is whether your issue is related to #680, a bug that I reported a while ago.

**Comment by reneeotten** [comment]
> I ran tests on Ubuntu 20.04. Note that the random number generator is seeded in the test module, but I guess this might still result in a different random number sequence? The only thing that then comes to my mind is whether your issue is related to #680, a bug that I reported a while ago.

I agree though with @newville and @Tillsten that this isn't really a bug, but originates from bad starting values as seems to be the case here as well. However, the tests should pass independent of the OS so we'll have to do something here.... either add less noise so that the initial guesses become better and the fit converges or loosen the tolerance.

**Comment by newville** [comment]
@reneeotten yes,  I can confirm that these tests are also failing for me on MacOS.    I apologize for merging this code.   

I believe the entire `check_guess_and_fit()` approach needs to be rewritten or scrapped.

**Pull request merged**